tests/h2-tests/tests/flow_control.rs

```diff
@@ -1561,3 +1561,55 @@ async fn data_padding() {
 
     join(srv, h2).await;
 }
+
+#[tokio::test]
+async fn poll_capacity_after_send_data_and_reserve() {
+    h2_support::trace_init!();
+    let (io, mut srv) = mock::new();
+
+    let srv = async move {
+        let settings = srv.assert_client_handshake_with_settings(frames::settings().initial_window_size(5)).await;
+        assert_default_settings!(settings);
+        srv.recv_frame(
+            frames::headers(1)
+                .request("POST", "https://www.example.com/")
+        )
+        .await;
+        srv.send_frame(frames::headers(1).response(200)).await;
+        srv.recv_frame(frames::data(1, &b"abcde"[..])).await;
+        srv.recv_frame(frames::data(1, &b""[..]).eos()).await;
+    };
+
+    let h2 = async move {
+        let (mut client, mut h2) = client::handshake(io).await.unwrap();
+        let request = Request::builder()
+            .method(Method::POST)
+            .uri("https://www.example.com/")
+            .body(())
+            .unwrap();
+
+        let (response, mut stream) = client.send_request(request, false).unwrap();
+
+        let response = h2.drive(response).await.unwrap();
+        assert_eq!(response.status(), StatusCode::OK);
+
+        stream.send_data("abcde".into(), false).unwrap();
+
+        stream.reserve_capacity(5);
+
+        // Initial window size was 5 so current capacity is 0 even if we just reserved.
+        assert_eq!(stream.capacity(), 0);
+
+        // FIXME(nox): This currently hangs because the capacity was reduced to 0 by the `send_data`
+        // call, but the stream's `send_capacity_inc` flag was never cleared and is still true,
+        // so `poll_for_capacity` immediately returns `Ready(Some(0))`.
+        let mut stream = h2.drive(util::wait_for_capacity(stream, 5)).await;
```

Comment:
@seanmonstar I managed to write a test that hangs on master, for the same reason your test hangs in my max send  buffer size branch. Do you think h2's innards are wrong, or is it `wait_for_capacity` that is wrong?

Reply:
https://github.com/hyperium/hyper/blob/6169db250c932dd012d391389826cd34833077b4/src/proto/h2/mod.rs#L129-L145

Hyper itself loops on `Ready(Some(0))`, so I guess `wait_for_capacity` is broken?

Reply:
Yes, I think `wait_for_capacity` is wrong to return `Pending` without having registered a task. It should probably be a loop, since calling `poll_capacity` again will have noticed `send_capacity_inc` is now false and registered for the next capacity wakeup.

Reply:
I changed `wait_for_capacity` in that commit then.